docs/config/01-configuration-file.md

```diff
@@ -67,6 +67,45 @@ require('ts-node').register({
 require('./karma.conf.ts');
 ```
 
+### Asynchronous Configuration
+
+The function exported by the configuration file may return a promise whose
+resolution will tell Karma when all configuration is complete, including the
+final call to `config.set()`.
+
+Just as with a synchronous configuration's return value not being used, an
+asynchronous configuration's resolved value will not be used. With both
+configuration types, `config.set()` MUST BE called to assign the configuration
+object.
+
+When using the command line interface, returning a promise is all that is
+needed.
+
+An example of a config that may need asynchronous support is a config that uses
+an async method to find an available port if the preferred port is already in
+use.
+
+```js
+// karma.conf.js
+const findAvailablePort = require('./findAvailablePort.js');
+
+module.exports = function configKarma(config) {
+  const myPreferredPort = 9876;
+  return findAvailablePort(myPreferredPort)
```

Comment:
This is a bad example IMHO, because it leaves the door wide open for race conditions to occur.

Reply:
The most common description of a race that I've had experience with is that multiple operations are being executed in parallel with no way to control the order of execution, which can lead to an unpredictable/unstable state.

I don't see the potential for race conditions here. The hypothetical `findAvailablePort` returns a promise. That promise must resolve before the onFulfilled parameter executes (which then execute **`config.set`**). 

As long as it is a `Promise` instance, or an object that is a conforming implementation of Promises/A+, then the order of operations is controlled and there is no race.

Are you concerned about promise rejections not getting handled in the example?

Reply:
I mean that getting a free port and then using that port. The port can be claimed by another process in the mean time. 

I've experienced this twice on the past.
1. On an old fashion buildserver that ran parallel builds.
2. During mutation testing with StrykerJS, where we also need to spin up multiple test runner processes.

The right way to choose a free port is by leaving it `undefined` and letting karma itself choose it.

Reply:
That is reasonable. In the time it takes between getting a port number and starting the server, something else could claim the port, especially if other async tasks happen before then.

The example can easily be changed to another process that is often asynchronous, "getThingFromFileSystem" or "getThingFromNetwork". I would rather the example draw attention to "hey, this is asynchronous" instead of "this is a way to solve this specific problem/task" even if the intent was for illustration. I'd even be ok with the super generic "doAsyncThing", as long as it gets the point across that it returns a promise.

Reply:
Exactly. I didn't see this PR at first and opened #3761 where I add documentation like that.